### PR TITLE
Statement and ConnectionState implementation

### PR DESCRIPTION
**Pull request by Kuleshovegor**
Basic Statement and ConnectionState implementation and new interfaces.

### CODE REVIEW

**Review by KirillKurdyukov — COMMENTED** [review]
Added Licence before package and rename kuleshovegor to Egor Kuleshov

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java [hunk 1]
```diff
@@ -23,24 +23,32 @@
...
-
-    private final TableClient tableClient;
+public class YdbConnection implements Connection {
+    private final YdbTypeResolver typeResolver;
+    private final R2dbcQueryParser r2dbcQueryParser;
 
     private volatile IsolationLevel isolationLevel = IsolationLevel.SERIALIZABLE;
     private volatile boolean autoCommit = true;
-    private volatile YDBConnectionState state;
+    private final YdbConnectionState state;
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/YdbConnection.java:46]: Why is state final? I have exec out transaction query, then start transaction.

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java [hunk 2]
```diff
@@ -23,24 +23,32 @@
...
 
-    public YDBConnection(TableClient tableClient) {
-        this.tableClient = tableClient;
+    public YdbConnection(YdbTypeResolver typeResolver, R2dbcQueryParser r2dbcQueryParser, YdbConnectionState state) {
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/YdbConnection.java:48]: Do you have R2dbcQueryParser is instance? Maybe is util class?

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java [hunk 3]
```diff
@@ -76,7 +84,12 @@ public Mono<Void> createSavepoint(String name) {
 
     @Override
     public Statement createStatement(String sql) {
-        return new YDBStatement(state, sql);
+        YdbQuery query =  r2dbcQueryParser.parseYdbQuery(sql);
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/YdbConnection.java:87]: good solution!

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java [hunk 4]
```diff
@@ -76,7 +84,12 @@ public Mono<Void> createSavepoint(String name) {
...
+
+        return switch (query.type()) {
+            case DATA -> new DataStatement(query, state, typeResolver);
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/YdbConnection.java:90]: Maybe DDL, DML enum? :)

File: src/main/java/tech/ydb/io/r2dbc/context/YdbConst.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package tech.ydb.io.r2dbc.context;
+
+/**
+ * @author kuleshovegor
+ */
+public final class YdbConst {
+    public static final String MISSING_VALUE_FOR_PARAMETER = "Missing value for parameter: ";
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/context/YdbConst.java:7]: Do we really want to store errors as constants?

File: src/main/java/tech/ydb/io/r2dbc/query/QueryType.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package tech.ydb.io.r2dbc.query;
+
+/**
+ * @author kuleshovegor
+ */
+public enum QueryType {
+    SCHEME,
+    DATA
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/query/QueryType.java:8]: DDL, DML

File: src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package tech.ydb.io.r2dbc.query;
+
+/**
+ * @author kuleshovegor
+ */
+public interface R2dbcQueryParser {
+    YdbQuery parseYdbQuery(String sql);
+}
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java:8]: Interfaces for a single class are not good.

File: src/main/java/tech/ydb/io/r2dbc/query/YdbExpression.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package tech.ydb.io.r2dbc.query;
+
+/**
+ * @author kuleshovegor
+ */
+public class YdbExpression {
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/query/YdbExpression.java:6]: Is enum?

Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/query/YdbExpression.java]: Deleted this class

File: src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java (hunk 7 above)
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java:26]: Made interface to class with static method.

File: src/main/java/tech/ydb/io/r2dbc/context/YdbConst.java (hunk 5 above)
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/context/YdbConst.java]: Used this message local

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java (hunk 1 above)
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/YdbConnection.java]: Revert to volatile.

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java [hunk 9]
```diff
@@ -23,24 +23,32 @@
...
+import tech.ydb.io.r2dbc.type.YdbTypeResolver;
 
 /**
  * @author Kirill Kurdyukov
  */
-public class YDBConnection implements Connection {
-
-    private final TableClient tableClient;
+public class YdbConnection implements Connection {
+    private final YdbTypeResolver typeResolver;
```
[the remark this answers is not in the document]
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/YdbConnection.java:41]: Made this class with static method.

File: src/main/java/tech/ydb/io/r2dbc/YdbConnectionFactory.java [hunk 10]
```diff
@@ -25,21 +25,21 @@
...
     private final TableClient tableClient;
 
-    public YDBConnectionFactory(TableClient tableClient) {
+    public YdbConnectionFactory(TableClient tableClient) {
         this.tableClient = tableClient;
     }
 
     @Override
     public Mono<? extends Connection> create() {
-        return Mono.just(new YDBConnection(tableClient));
```
[the remark this answers is not in the document]
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/YdbConnectionFactory.java:38]: This is temporary for building a project.

File: src/main/java/tech/ydb/io/r2dbc/YdbBatch.java [hunk 11]
```diff
@@ -45,6 +46,6 @@ public Batch add(String sql) {
 
     @Override
     public Publisher<? extends Result> execute() {
-        return new YDBStatement(state, String.join(";", statements)).execute();
+        return Mono.empty();
```
[the remark this answers is not in the document]
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/YdbBatch.java:49]: This is temporary for building a project

File: src/main/java/tech/ydb/io/r2dbc/statement/SchemaStatement.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import tech.ydb.io.r2dbc.query.YdbQuery;
+import tech.ydb.io.r2dbc.state.YdbConnectionState;
+import tech.ydb.io.r2dbc.type.YdbTypeResolver;
+
+/**
+ * @author kuleshovegor
+ */
+public class SchemaStatement extends BaseStatement {
```
[the remark this answers is not in the document]
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/statement/SchemaStatement.java]: Renamed to DMLYdbStatement.

File: src/main/java/tech/ydb/io/r2dbc/statement/ScanStatement.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import tech.ydb.io.r2dbc.query.YdbQuery;
+import tech.ydb.io.r2dbc.state.YdbConnectionState;
+import tech.ydb.io.r2dbc.type.YdbTypeResolver;
+
+/**
+ * @author kuleshovegor
+ */
+public class ScanStatement extends BaseStatement {
```
[the remark this answers is not in the document]
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/statement/ScanStatement.java]: Deleted.

File: src/main/java/tech/ydb/io/r2dbc/statement/DataStatement.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import tech.ydb.io.r2dbc.query.YdbQuery;
+import tech.ydb.io.r2dbc.state.YdbConnectionState;
+import tech.ydb.io.r2dbc.type.YdbTypeResolver;
+
+/**
+ * @author kuleshovegor
+ */
+public class DataStatement extends BaseStatement {
```
[the remark this answers is not in the document]
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/statement/DataStatement.java]: Renamed to DMLYdbStatement.

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java (hunk 3 above)
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/YdbConnection.java]: Is it issue?

File: src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tech.ydb.io.r2dbc.query;
+
+/**
+ * @author Egor Kuleshov
+ */
+public class R2dbcQueryParser {
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java:22]: Rename to YdbSqlParser

R2dbc prefix here is excess

File: src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    public static YdbQuery parseYdbQuery(String sql) {
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/query/R2dbcQueryParser.java:23]: public static YdbQuery parse(String sql) {

YdbQuery is result

File: src/main/java/tech/ydb/io/r2dbc/query/YdbQuery.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+                    }
+
+                    String prmType = values.get(prm).getType().toString();
+                    yql.append("DECLARE ")
+                            .append(prm)
+                            .append(" AS ")
+                            .append(prmType)
+                            .append(";\n");
+
+
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/query/YdbQuery.java:60]: delete whitespaces

File: src/main/java/tech/ydb/io/r2dbc/statement/DDLYdbStatement.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+import io.r2dbc.spi.Statement;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import tech.ydb.io.r2dbc.query.YdbQuery;
+import tech.ydb.io.r2dbc.state.YdbConnectionState;
+
+/**
+ * @author Egor Kuleshov
+ */
+public class DDLYdbStatement extends YdbStatement {
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/statement/DDLYdbStatement.java:30]: YdbDDLStatement

File: src/main/java/tech/ydb/io/r2dbc/statement/binding/Bindings.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package tech.ydb.io.r2dbc.statement.binding;
+
+/**
+ * @author Egor Kuleshov
+ */
+public interface Bindings extends Iterable<Binding> {
+    Binding getCurrent();
+    void add(Binding binding);
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/statement/binding/Bindings.java:24]:
```
    Binding getCurrent();
    
    void add(Binding binding);
```

File: src/main/java/tech/ydb/io/r2dbc/state/YdbConnectionState.java [hunk 20]
```diff
@@ -16,15 +16,17 @@
...
 /**
  * @author Kirill Kurdyukov
  */
-public sealed interface YDBConnectionState permits Close, InTransaction, OutTransaction {
+public sealed interface YdbConnectionState permits Close, InTransaction, OutTransaction {
 
-    CompletableFuture<Result<DataQueryResult>> executeDataQuery(String yql, Params params);
+    Mono<YdbDMLResult> executeDataQuery(String yql, Params params);
+
+    Mono<YdbDDLResult> executeSchemaQuery(String yql, Params params);
```
[the remark this answers is not in the document]
Comment (Kuleshovegor) [reply src/main/java/tech/ydb/io/r2dbc/state/YdbConnectionState.java]: Why? Mono should be parameterized

File: src/main/java/tech/ydb/io/r2dbc/state/YdbConnectionState.java [hunk 21]
```diff
@@ -16,15 +16,17 @@
...
-import tech.ydb.table.query.DataQueryResult;
+import reactor.core.publisher.Mono;
+import tech.ydb.io.r2dbc.result.YdbDMLResult;
+import tech.ydb.io.r2dbc.result.YdbDDLResult;
 import tech.ydb.table.query.Params;
 
 /**
  * @author Kirill Kurdyukov
  */
-public sealed interface YDBConnectionState permits Close, InTransaction, OutTransaction {
```
Comment (Kuleshovegor) [inline src/main/java/tech/ydb/io/r2dbc/state/YdbConnectionState.java:27]: Removed sealed for test :( Mockito unsupported seald

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java [hunk 22]
```diff
@@ -23,24 +23,30 @@
...
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/YdbConnection.java:41]: delete fiels, it is util class

File: src/main/java/tech/ydb/io/r2dbc/YdbConnection.java [hunk 23]
```diff
@@ -23,24 +23,30 @@
...
 
     private volatile IsolationLevel isolationLevel = IsolationLevel.SERIALIZABLE;
     private volatile boolean autoCommit = true;
-    private volatile YDBConnectionState state;
+    private volatile YdbConnectionState state;
 
-    public YDBConnection(TableClient tableClient) {
-        this.tableClient = tableClient;
+    public YdbConnection(YdbTypeResolver typeResolver, YdbConnectionState state) {
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/YdbConnection.java:47]: public YdbConnection(YdbConnectionState state) {

File: src/main/java/tech/ydb/io/r2dbc/query/YdbQuery.java [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    YdbQuery(String yql, List<String> indexesArgsNames, QueryType queryType) {
+        this.yqlQuery = yql;
+        this.indexesArgsNames = indexesArgsNames;
+        this.type = queryType;
+    }
+
+    public String getYqlQuery(Params params) throws SQLException {
+        StringBuilder yql = new StringBuilder();
+
+        if (indexesArgsNames != null) {
```
Comment (KirillKurdyukov) [inline src/main/java/tech/ydb/io/r2dbc/query/YdbQuery.java:45]: Maybe indexesArgsNames is not null param?

**Review by KirillKurdyukov — APPROVED** [review]

**Pull request merged**